docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,389 @@
+- Provide an update strategy without draining the node.
+- Provide in-place updates only for the Kubernetes version in worker pools where the OS does not support in-place updates.
+
+## Proposal
+
+In short, Gardener should support new update strategies that do not require deleting and recreating the node when the Kubernetes minor or Machine image version is updated.
+
+### Approach
+
+The `machine-controller-manager` and `gardener-node-agent` components will be used to perform in-place updates. `machine-controller-manager` will provide an interface that prepares the machine/node for an update. This interface can be used by other components to determine if a node is ready for an update, allowing the other component to perform the update on the node and then notify the `machine-controller-manager` that the node has been successfully updated. _In the case of Gardener, this component will be the `gardener-node-agent`._
+
+This design was chosen to leverage the `machine-controller-manager`'s node rollout orchestration functionality while ensuring that it remains a standalone component.
+
+### Prerequisites
+
+The only prerequisite is that the machine image/OS on the node must support in-place updates and provide a tool or utility to initiate the OS update. This update tool/utility should meet the following requirements:
+
+1. **Failure Recovery**: In case of a failure, the OS should be able to revert to the previous version.
+2. **Retry Configuration**: Support for configuring the number of update retries.
+3. **Registry Configuration**: Ability to specify the registry from which to pull the OS image.
+
+### Update Strategies
+
+Gardener will introduce two additional update strategies, `InPlaceUpdate` and `InPlaceUpdateOnLabel`, for worker pools, in addition to the existing `RollingUpdate`. The two new strategies are described as follows:
+
+#### InplaceUpdate strategy
+
+1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the changed worker pool with the `candidate-for-update` label. This label is used to identify nodes that require an update.
+2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.
+3. `machine-controller-manager` will select as many nodes as are permitted by the `maxUnavailable` setting for the worker pool and label it/them with `selected-for-update`. This label can be used to identify nodes that are about to or are currently undergoing an update.
+4. `machine-controller-manager` will cordon and drain the node/nodes and label it/them with the `ready-for-update` label once the drain has completed.
+5. `gardener-node-agent` will detect the `ready-for-update` label and perform the update on the machine. It will wait for the update to complete (with a specified timeout).
```

Comment:
There will be two timeout inplace - 
- GNA - if update is not able to get applied in a specific timeout, it will consider it as failed update and add the label update-failed the node, so MCM knows the update failed.
- MCM - This timeout if for the case when GNA is not able to report back for ex - in case of OS failed to boot, if this timeout is reached, MCM will consider it as a failed update.